utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+//TODO: remove once the functions are used.
+#![allow(dead_code)]
+use crate::{Error, TxUndo};
+use common::chain::{OutPoint, OutPointSourceId, Transaction, TxOutput};
+use common::primitives::{BlockHeight, Id, Idable};
+use std::collections::BTreeMap;
+use std::fmt::{Debug, Formatter};
+
+use common::chain::block::Block;
+use parity_scale_codec::{Decode, Encode};
+
+pub mod utxo_storage;
+
+//todo: proper placement and derivation of this max
+const MAX_OUTPUTS_PER_BLOCK: u32 = 500;
+
+// Determines whether the utxo is for the blockchain of for mempool
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub enum UtxoSource {
+    /// At which height this containing tx was included in the active block chain
+    BlockChain(BlockHeight),
+    MemPool,
+}
+
+impl UtxoSource {
+    fn is_mempool(&self) -> bool {
+        match self {
+            UtxoSource::BlockChain(_) => false,
+            UtxoSource::MemPool => true,
+        }
+    }
+
+    fn blockchain_height(&self) -> Result<BlockHeight, Error> {
+        match self {
+            UtxoSource::BlockChain(h) => Ok(*h),
+            UtxoSource::MemPool => Err(crate::Error::NoBlockchainHeightFound),
+        }
+    }
+}
+
+/// The Unspent Transaction Output
+#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+pub struct Utxo {
+    output: TxOutput,
+    is_block_reward: bool,
+    /// identifies whether the utxo is for the blockchain or for mempool.
+    source: UtxoSource,
+}
+
+impl Utxo {
+    pub fn new(output: TxOutput, is_block_reward: bool, height: BlockHeight) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            source: UtxoSource::BlockChain(height),
+        }
+    }
+
+    /// a utxo for mempool, that does not need the block height.
+    pub fn new_for_mempool(output: TxOutput, is_block_reward: bool) -> Self {
+        Self {
+            output,
+            is_block_reward,
+            source: UtxoSource::MemPool,
+        }
+    }
+
+    pub fn is_block_reward(&self) -> bool {
+        self.is_block_reward
+    }
+
+    pub fn height(&self) -> &UtxoSource {
+        &self.source
+    }
+
+    pub fn output(&self) -> &TxOutput {
+        &self.output
+    }
+
+    pub fn set_height(&mut self, value: UtxoSource) {
+        self.source = value
+    }
+}
+
+pub trait UtxosView {
+    /// Retrieves utxo.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
+
+    /// Checks whether outpoint is unspent.
+    fn has_utxo(&self, outpoint: &OutPoint) -> bool;
+
+    /// Retrieves the block hash of the best block in this view
+    fn get_best_block_hash(&self) -> Option<Id<Block>>;
+
+    /// Estimated size of the whole view (None if not implemented)
+    fn estimated_size(&self) -> Option<usize>;
```

Comment:
I'm not sure what you're referring to and why Some(0) is interesting at all. If this is about why we don't use 0 as None, it's because I wanna make sure the developer handles that case.

Reply:
IIRC previously the value `0` was used to denote that it's not implemented which I preferred. I think a better question is that can the estimated size of the view be zero if it's implemented? If it can be and it's totally valid value for the view then there's no problem. If it can't be zero if an implementation exists, then `Some(0)` is an unnecessary special case we need to be mindful of. Or use `Option<NonZeroUsize>`.

Reply:
This depends primarily on the goal of doing that. If your goal is to optimize for byte size and utilize all the space there is (like in MCUs), then sure, using just a usize will be sufficient. However, my goal is different. My goal is to force the developer to handle all cases, 100% at compile-time. When using an int for two meanings, it's very easy to overlook the case when it's not implemented. If None and 0 are equivalent for a certain caller, a developer can easily do a `unwrap_or(0)`.

Reply:
I'm not thinking about optimizations but I am confused because previously, if I remember correctly, `0` was used to denote that implementation doesn't exist which implies that when this estimation has been implemented it cannot return zero (because that would be an error?). Is this a correct assessment/was it ever true?

My point is that if an implementation cannot have a size of zero when it exists, wrapping the return value in an `Option<T>` doesn't make things easier because now you have to match `None`, `Some(0)` (two possible errors) and `Some(_)` which doesn't seem easier to me. I'm not talking about optimizations but I'm hung up on this zero return value.

Reply:
> was it ever true

Yes it was, and it's like that in Bitcoin. I changed it.

> doesn't make things easier because now you have to match

That's exactly the point. "Easier" is not really the goal here. The goal here is "correct". How can we make it correct? We force the developer to handle all possible cases.

Some(0) is not really the interface's problem. It's an implementation problem. I think this is the cause of the confusion. You're mixing implementation issues and interface design. The right answer there is that you test and ensure that your implementation doesn't yield Some(0) through unit tests. But you can't guarantee that all developers who use this interface will handle the 0 case unless you build it in the interface.

This is just like Eq and PartialEq. There's a reason that distinction exists in the interface. PartialEq basically forces you to handle a corner case when NaN != NaN. Here also we want to force the dev to handle a corner case, unlike C/C++ where the dev can just omit this corner case and then when it happens, bad things happen.

Reply:
Ok so in that case `Some(0)` is never a valid value which is my problem here. Why not use `Option<NonZeroUsize>` in that case? Then the interface makes more sense: either you get `Some(_)` and you know it's valid or you get `None` which means it's not implemented. That's also the safer option because now there's no way to ignore the zero return value because you never get that.

Reply:
That's alright. We can change it to that. My only contention there is that we're probably using too much from the outside. But since it's std::num crate, it's not a big deal.

Reply:
Actually I just noticed that the estimated size can be zero if the view is empty... so 0 is valid. I thought from the first look this is about a UtxoEntry, not a UtxoView.